[Merged by Bors] - feat(src/ring_theory/algebra): define equivalence of algebras

ocfnash opened the pull request.

Commit:
feat(src/ring_theory/algebra): define equivalence of algebras

urkud commented on src/ring_theory/algebra.lean:
```diff
@@ -335,6 +337,51 @@ ext $ λ x, show φ₁.to_linear_map x = φ₂.to_linear_map x, by rw H
+
+instance has_coe_to_ring_equiv : has_coe (A₁ ≃ₐ[R] A₂) (A₁ ≃+* A₂) := ⟨alg_equiv.to_ring_equiv⟩
+
+@[norm_cast] lemma coe_ring_equiv (e : A₁ ≃ₐ[R] A₂) (a : A₁) : (e : A₁ ≃+* A₂) a = e a := rfl
```
Please use `((e : A₁ ≃+* A₂) : A₁ → A₂) = e`.

Commit:
Feedback from CR

urkud commented on src/ring_theory/algebra.lean:
```diff
@@ -335,6 +337,51 @@ ext $ λ x, show φ₁.to_linear_map x = φ₂.to_linear_map x, by rw H
+
+instance has_coe_to_ring_equiv : has_coe (A₁ ≃ₐ[R] A₂) (A₁ ≃+* A₂) := ⟨alg_equiv.to_ring_equiv⟩
+
+@[norm_cast] lemma coe_ring_equiv (e : A₁ ≃ₐ[R] A₂) : ((e : A₁ ≃+* A₂) : A₁ → A₂) = e := rfl
```
```suggestion
@[simp, norm_cast] lemma coe_ring_equiv (e : A₁ ≃ₐ[R] A₂) : ((e : A₁ ≃+* A₂) : A₁ → A₂) = e := rfl
```

Commit:
Update src/ring_theory/algebra.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

Commit:
Add additional lemmas suggested by Yury

Note that I have also changed the definition of `alg_equiv` so that it now
also extends `equiv`, `mul_equiv`, `add_equiv`. This adds no arguments to
the constructor since `ring_equiv` already extends these but it saves some
indirection in a few places.

Commit:
Add missing nolint statements

urkud commented on src/ring_theory/algebra.lean:
```diff
@@ -335,6 +337,94 @@ ext $ λ x, show φ₁.to_linear_map x = φ₂.to_linear_map x, by rw H
+  [ring A₁] [ring A₂] [algebra R A₁] [algebra R A₂] (e : A₁ ≃ₐ[R] A₂) :
+  ∀ x y, e (x - y) = e x - e y := e.to_add_equiv.map_sub
+
+instance has_coe_to_alg_equiv : has_coe (A₁ ≃ₐ[R] A₂) (A₁ →ₐ[R] A₂) :=
```
```suggestion
instance has_coe_to_alg_hom : has_coe (A₁ ≃ₐ[R] A₂) (A₁ →ₐ[R] A₂) :=
```

ocfnash replied:
🤦, thanks!

urkud commented on src/ring_theory/algebra.lean:
```diff
@@ -335,6 +337,94 @@ ext $ λ x, show φ₁.to_linear_map x = φ₂.to_linear_map x, by rw H
+  rfl
+
+lemma injective (e : A₁ ≃ₐ[R] A₂) : function.injective e :=
+  function.injective_of_left_inverse e.left_inv
```
Does `e.to_equiv.injective` work? This way it's clear that we export something from `to_equiv`, not proving something new. Same for `surjective` and `bijective`.

ocfnash replied:
Does indeed work, thanks.

urkud committed:
Update src/ring_theory/algebra.lean

Commit:
Minor simplifications / golfing
